### PR TITLE
manually check for safe downcasting

### PR DESCRIPTION
**Pull request by dschult**
This simply implements @stefanv 's suggestion for how to manually check that downcasting is safe.

I haven't added any tests, but I think this should do what we want as far as downcasting but spending a little CPU time to check if there will be a problem downcasting. If that cost is too high it seems like we should downcast aggressively like was done before.

### CODE REVIEW

**Review by stefanv — APPROVED** [review]

File: scipy/sparse/linalg/_dsolve/linsolve.py [hunk 1]
```diff
@@ -404,6 +404,12 @@ def csc_construct_func(*a, cls=type(A)):
     if (M != N):
         raise ValueError("can only factor square matrices")  # is this true?
 
+    # check for safe downcasting
+    ii = np.iinfo(np.int32)
+    # check indices because they should be larger than indptr
+    if np.any(A.indices > ii.max || A.indices < ii.min):
```
Comment (perimosocordiae) [inline scipy/sparse/linalg/_dsolve/linsolve.py]: We only need to check the max side here, as indices are nonnegative. Can you update this to do something like what I wrote in https://github.com/scipy/scipy/pull/18644#issuecomment-1594822601 ?

We should probably pull that out into a helper function, too, like: `indices, indptr = _safe_cast_indices(A)`

Comment (dschult) [reply scipy/sparse/linalg/_dsolve/linsolve.py]: I have pulled out the code for checking safe downcasting and changed it to follow @perimosocordiae ‘s suggested faster checks. Thoughts?

**Review by stefanv — APPROVED** [review]

**Comment by stefanv** [comment]
Looks good, thanks for adding the short circuits!

**Comment by perimosocordiae** [comment]
Thanks, merged!

**Pull request merged**